### ISSUE TITLE
New Feature: Date Picker

### ISSUE THREAD

**Issue by TedGoas** [StackExchange/Stacks#188]
**Is your feature request related to a problem? Please describe.**
Stack Overflow has a few instances of a data picker. The current solutions are all different (jQuery UI) and not very robust.

**Describe the solution you'd like**
A date picker in Stacks that includes:
1. A few predefined ranges.
2. Support for custom date ranges.
3. Theming support (eg. Enterprise, network sites)

**Describe alternatives you've considered**
jQuery

**Additional context**
@UXPiper worked on a few mockups [here](https://stackexchange.invisionapp.com/share/XUHRLFJBQTG#/screens/322843077) and [here](https://stackexchange.invisionapp.com/share/XUHRLFJBQTG#/screens/308162458):

<img width="1037" alt="screen shot 2018-11-19 at 4 11 07 pm" src="https://user-images.githubusercontent.com/1172461/48735220-c0540400-ec15-11e8-93c1-3d49c4903d12.png">

**Comment by UXPiper** [comment]
This is the final design spec: https://stackexchange.invisionapp.com/share/MTP7KOWEZ7R#/332358366_SO-Stacks--v01__Design-Spec__Date-Picker

**Comment by dancormier** [comment]
[PLATENG-734](https://stackoverflow.atlassian.net/browse/PLATENG-734)

### UPDATE
I linked to a Jira issue and decided to put my notes there. That doesn't make much sense considering this is a Stacks component discussion we're having.

The next few comments are my notes (and some helpful input from @bnickel) ported from the Jira issue.

**Comment by dancormier** [comment]
From Jira, [@dancormier commented 5 days ago](https://stackoverflow.atlassian.net/browse/PLATENG-734?focusedCommentId=202960):

---
# [vanillajs-datepicker](https://github.com/mymth/vanillajs-datepicker)

A quick search lead me to [vanillajs-datepicker](https://github.com/mymth/vanillajs-datepicker), which looks pretty badass* overall.

*badass in this context means accessible, configurable, and full featured.

## Size

I did a back-of-the-napkin calculation and found the minified CSS + JS to land around 10-11KB†

†I need to verify this and this may depend on which features we’d like to include in our component.

## Features

- [Roll your own element classes](https://mymth.github.io/vanillajs-datepicker/#/?id=other-frameworks). This would allow us to apply Stacks atomic classes.
- Includes [options for displaying inline or popover](https://mymth.github.io/vanillajs-datepicker/#/overview?id=types).
- Includes option for [individual date, range selection](https://mymth.github.io/vanillajs-datepicker/#/overview?id=date-range-picker), and [multi-date selection](https://mymth.github.io/vanillajs-datepicker/#/overview?id=multidate-mode).

## Maintenance 

- It’s being actively maintained, albeit with the vast majority of contributions [from a single contributor](https://github.com/mymth/vanillajs-datepicker/graphs/contributors) (maybe a small OSS sponsorship candidate).

## Potential issues

- Single maintainer.
- Expects an `<input type=”text” />`. Seems like a `type=”date”` would make more sense (I assume it’s to avoid fighting with native date pickers).
- Accessibility concerns.
    - Yellow flag: I noticed it uses `<span>`s for the date items. Feels like that should be `<button>`s.
    - [Keyboard navigation](https://mymth.github.io/vanillajs-datepicker/#/overview?id=keyboard-operation) felt pretty intuitive, but not sure it cuts it.
- What’s our bundle size tolerance here?

**Comment by dancormier** [comment]
From Jira, [@dancormier commented 5 days ago](https://stackoverflow.atlassian.net/browse/PLATENG-734?focusedCommentId=202968):

---

# [flatpickr](https://github.com/flatpickr/flatpickr)

## Features

- There’s an actively maintained [Stimulus.js wrapper](https://github.com/adrienpoly/stimulus-flatpickr), along with wrappers from other frameworks.
- [Turns into a native input on mobile](https://flatpickr.js.org/mobile-support/).

## Maintenance

- Latest release came in December 2020, but the repo has had many contributions since then. 

## Potential issues

- Far from headless. UI doesn’t seem to be all that customizable (can’t add custom classes where ever).
- [Lots of issues on GitHub](https://github.com/flatpickr/flatpickr/issues).

## Investigate

- Size.
- Is there some way we can make this look like Stacks?
- Additional killer features? I haven’t dug hard into this one.

**Comment by dancormier** [comment]
From Jira, [@bnickel commented 5 days ago](https://stackoverflow.atlassian.net/browse/PLATENG-734?focusedCommentId=202981):

---

For awareness, here’s some places where we use date pickers today and ones with nuance.  We generally, but not always, deal with date ranges.

## https://stackoverflow.com/admin/community-events 

![20210802-135016](https://user-images.githubusercontent.com/647177/127903279-d455ce39-18bf-4d45-bfa3-7b078e7a0f63.png)

This is the only scenario I know of where time is involved.  Also the only one where we display in local timezone.

## https://stackoverflow.com/admin/user-activity 

![20210802-135121](https://user-images.githubusercontent.com/647177/127903450-94de0e66-73a5-4901-8f4c-7bf95615d201.png)

## https://stackoverflow.com/admin/analytics 

![20210802-135129](https://user-images.githubusercontent.com/647177/127903468-52476f57-8b8a-4f32-b13a-3bd524bb194f.png)

## https://stackoverflow.com/ads/sponsorships/add 

![20210802-135138](https://user-images.githubusercontent.com/647177/127903480-064aa3f1-ddae-45c8-9e58-93f3b1edf0b1.png)

## https://stackoverflow.com/cm/elections/add 

![20210802-135142](https://user-images.githubusercontent.com/647177/127903503-79db864b-1af8-42d7-8539-c7b403fd4ad7.png)

This one only lets you pick Tuesdays.

## https://stackoverflow.com/events/core 

![20210802-135146](https://user-images.githubusercontent.com/647177/127903520-7f190c89-1f5b-4c75-990c-3dc8424dd767.png)

This one uses the native date input.

## Channels admin

![20210802-135151](https://user-images.githubusercontent.com/647177/127903547-638583fc-973f-400b-b92f-f60b6bccd833.png)
![20210802-135154](https://user-images.githubusercontent.com/647177/127903560-d179cfb4-92c8-4214-9d0c-e6c6de76b754.png)

These ones support an unbounded date.

## Enterprise

Enterprise has some date pickers and I believe a custom component, but I’ve not looked into those.

---

Then I [thanked Brian](https://stackoverflow.atlassian.net/browse/PLATENG-734?focusedCommentId=202982).

**Comment by dancormier** [comment]
From Jira, [@dancormier commented 3 days ago](https://stackoverflow.atlassian.net/browse/PLATENG-734?focusedCommentId=203488):

---

# `<input type="date">`

So, yeah we might be better off starting with a good, old fashioned `<input type="date">`. Seems like it would do the job when it comes to most of the date inputs we already have littered across the site.

If we went this route, we could _start_ with this native input and bring in a different solution when it’s needed. I’d consider this the YAGNI approach.

## Handy references

- [`<input type="date">` - HTML: HyperText Markup Language | MDN](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/date)
- [Date and time input types | Can I use... Support tables for HTML5, CSS3, etc](https://caniuse.com/input-datetime)

## Features

- It’s a damn native input 🎉 It feels right for your platform and would probably be a good choice when considering [Jakob’s law](https://lawsofux.com/jakobs-law/).
- No JS (though we could lightly extend it with JS). You can’t get lighter-weight than this.

## Potential issues

- That damn calendar icon is on Chrome, Edge, and Opera. It would need a `filter: invert` for dark mode.
    - [Change color of Chrome's calendar icon in HTML Date Input](https://stackoverflow.com/a/62162805/652353)
    - And no, we shouldn’t just `display: none`; on it because it’s the only way to trigger the calendar widget.
- Calendar widget is not embeddable in-page
- Date range isn’t the cleanest experience (requires two separate inputs).
- I’ve seen accessibility issues related to date inputs specifically when using voice-based assistive technologies. I’m not sure of the situation there _currently_, but I’d note that this input is almost certainly more accessible to all js-based date pickers.
    - [input(type-date)_element (html) | Accessibility Support](https://a11ysupport.io/tech/html/input(type-date)_element)
    - [Is input type="date" ready for use in accessible websites? - Hassell Inclusion](https://www.hassellinclusion.com/blog/input-type-date-ready-for-use/)
- While the input supports `min` and `max` attributes, they seems to only limit selection on the calendar widget the input provides in some cases. Keyboard input can get around those attributes.
    - Observed in Firefox and Safari. Not an issue in Chrome AFAICT.

**Comment by aaronshekey** [comment]
Thanks for gathering this info, @dancormier. Seems like we could refactor to native date inputs on the way to something more fully-featured / customizable. Leaning toward Vanilla?

**Comment by dancormier** [comment]
@aaronshekey yep, leaning vanilla for sure. I think it will meet _most_ of our needs and can be extended if we absolutely need it*. If we're okay with working with a simplified picker to start, vanilla is probably the flavor we want.

*I'm thinking in the case of the [one date picker at `/cm/elections/add` that only allows Tuesdays](https://github.com/StackExchange/Stacks/issues/188#issuecomment-891217879).

**Comment by dancormier** [comment]
**TL;DR**: I advise we implement the date picker without third party libraries and stick with a native `<input type="date" />`.

# Requirements and goals

Based mostly on the existing implementations provided by @bnickel, it seems like our requirements largely overlap from one implementation to another.

## Common requirements

### Must have

- A input field to type the date
- A calendar interface to select a date
- An accessible interface
- Minimum [browser support](https://browsers.stackoverflow.design/)
- Actively maintained

### Nice to have

- Small (or no) JS footprint
- Dark mode support
- Simple localization
- Date range support

### One-off requirements

Here are some features implemented in individual existing date pickers on our products:

- Limit date selection to specific day(s) of week
    - See [comment on https://stackoverflow.com/cm/elections/add](https://github.com/StackExchange/Stacks/issues/188#issuecomment-891217879)
- Conjoined date + time selection*
    - See [comment on https://stackoverflow.com/ads/sponsorships/add](https://github.com/StackExchange/Stacks/issues/188#issuecomment-891217879)

# Findings and recommendations

I'd advise replacing current date picker implementations with `<input type="date">` in the interum (at least). This way, we can provide the user with a consistent, dependable, lightweight date picker and extend (or replace) it at a later date if the need arises.

---

## Clear advantages

Using a native input with no third party libraries has clear advantages over third party implementations:

- **Developer and designer effort**: Because it's a native input, it'll use the native styling defined by the browser and OS. This will ensure it feels natural to the user (at the cost of not being able to style part such as the calendar).
- **Size**: Doesn't get much lighter than a single DOM element and a little CSS. We can augment it with JS, but I suspect we'll often be fine with no JS.
- **Browser support**: It's supported in all browsers we support.
- **Maintenance**: It's unlikely to be deprecated anytime soon and we'd have a big heads up if it was.

## Likely advantages

It seems a native input will excel at:

- **Accessibility**: I'm confident this will be the most accessible option, but I've read(citation) some critiques of the native input when used in conjunction with VoiceOver.
- **Localization**: I've had trouble simulating this native input adhering to local date formats, but I suspect it does. I'd love some confirmation.

## Disadvantages/compromises

There are a few compromises if we go with a native input:

- **Customization/theming**: We won't be able to customize the calendar interface and the calendar icon shown in some browsers.
- **Selection of arbitrary dates**: We could prevent the user from submitting a given date, but we won't be able to disable dates on the calendar picker arbitrarily, like in the current picker at [https://stackoverflow.com/cm/elections/add](https://github.com/StackExchange/Stacks/issues/188#issuecomment-891217879).
when it comes to developer/design effort, size, and browser support. It's likely to provide the most accessible interface (though issues have been reported).
- **Dark mode**: The input can have light/dark styling applied like most other inputs, but the calendar seems to stay white no matter the dark/light context.
- **Date range support**: The [first comment on this issue](https://github.com/StackExchange/Stacks/issues/188#issue-382394663) asks for "support for custom date ranges" and "a few predefined ranges", which this solution cannot provide.
- **Conjoined date time picker**: See "**Including time selection in the date picker**" below*

## *Including time selection in the date picker

Many third-party JS date pickers include an option add time selection to the date input. If we go the vanilla route, we might be able to get away with the picker _just_ being HTML and CSS. We have a couple options if we go the vanilla route:

### Using `<input type="datetime-local" />`

This could be an input type we support and endorse in addition to `date`. At the time of this writing, it's supported in all  browsers Stacks supports _except_ Firefox... BUT, after digging into [a 5 year old bug report](https://bugzilla.mozilla.org/show_bug.cgi?id=1283388), it looks like Firefox support is on it's way (as of 2 days ago 😮).

See also [Date and time input types | Can I use...](https://caniuse.com/input-datetime)

### Using separate `<input type="date" />` and `<input type="time" />` inputs

This is probably the best vanilla option _initially_. It would allow us to replace existing date and time pickers with a lightweight, well-supported option.
